### PR TITLE
Make `network-mux` independent of `typed-protocols`

### PR DESCRIPTION
**Pull request by coot**
This PR makes `network-mux` indenependent of `typed-protocols`.  In tests it is using a simple implementation of `ReqResp` protocol without `typed-protocols`.  I initially thought it will be easier, but as a side effect it made quite explicit that `network-mux` depends on using an incremental decoder. (I am also fine with using typed-protocols in the tests, dropping one commit from this PR).

This has some consequences on `ouroboros-network`, suggestions for better names in `Ouroboros.Network.Mux` are welcome.   One suggestion I have is to move `Ouroboros.Network.Mux` as `Ouroboros.Network` (drop the `Mux` prefix in names), and possibly include more useful exports there e.g. (`Network.Mux.Interface`, `Ouroboros.Network.NodeToClient` and `Ouroboros.Network.NodeToNode`, and possible some other stuff). This could go into a separate PR though.

### CODE REVIEW

File: network-mux/test/Test/Mux/ReqResp.hs [hunk 1]
```diff
@@ -0,0 +1,202 @@
+{-# LANGUAGE GADTs               #-}
+{-# LANGUAGE NamedFieldPuns      #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE RankNTypes          #-}
+
+-- | A simple ReqResp protocol, togetether with an incremental decoder.  We need
+-- Mux is chopping messages into 'MuxSDU's of a fixed size.  We use incremental
```
Comment (karknu) [inline network-mux/test/Test/Mux/ReqResp.hs]: Odd grammar.

File: network-mux/test/Test/Mux/ReqResp.hs [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+    go Nothing
+  where
+    go :: Maybe LBS.ByteString
+       -> ReqRespServer req resp m a
+       -> m a
+    go trailing ReqRespServer {recvMsgReq, recvMsgDone} = do
+      res <- withLiftST $ \liftST -> runDecoderWithChannel liftST channel trailing decode
+
+      case res of
+        Left err -> error $ "runServer: deserialise error " ++ show err
```
Comment (karknu) [inline network-mux/test/Test/Mux/ReqResp.hs]: Would be good if we ran `traceWith` for the deserialise error too.

Comment (coot) [reply network-mux/test/Test/Mux/ReqResp.hs]: Good point, I all add it.

Comment (coot) [reply network-mux/test/Test/Mux/ReqResp.hs]: @karknu there's no way to do that. The reason is that `TraceSendRecv` now only allows to put trace messages not decoding errors, but I think we should extend it to something like:
```
data TraceSendRecv msg st failure
  = TraceSend msg
  | TraceRecv msg
  | TraceDecodingError st failure
  deriving Show
```
The `TraceDecodingError` could trace the current protocol state & the codec failure.  But this is out side of this PR.

Comment (coot) [reply network-mux/test/Test/Mux/ReqResp.hs]: Actually I am wrong, this `TraceSendRecv` is private here, so we can change it as we see it fit in this PR.

Comment (coot) [reply network-mux/test/Test/Mux/ReqResp.hs]: But extending the original trace is also a good idea.

Comment (karknu) [reply network-mux/test/Test/Mux/ReqResp.hs]: Then I think you should, and we can file an issue to add it to `typed-protocol` too.

Comment (coot) [reply network-mux/test/Test/Mux/ReqResp.hs]: Done, yes adding an issue is a good idea.

**Pull request merged**